### DISCUSSION TITLE
Поделитесь пожалуйста опытом подбора настроек под себя.

### DISCUSSION THREAD

**Discussion by wassalam01** [hufrea/byedpi#224] [General]
Огромное спасибо автору за проделанную работу. Моя ситуация осложнена еще и тем что у меня мак, но не в этом суть вопроса.
А суть вопроса следующая  она заключается в моем опыте использования как пользователя. Я скачал и запустил программу само-собой ничего не заработало, долго отдуплял запустилась ли она, или нет. Потом понял что запустилась, полез разбираться с параметрами их много они классные, только не все очевидны для меня, нашел в ишью чужие настройки которые сработали, потом очередная волна жалоб на то что ютуб перестал работать, экспериментирую с настройками завышаю цифры, бац работает радуюсь (не долго) какой я мамкин хакер. Но я не понимаю что я делаю, в этом и суть вопроса, понимаю что за 21 день можно выучить си, но у меня не получилось ) 
Вопрос, как можно понимать какие крутилки крутить, что мне прочитать для этого. Смотреть TCP дампом пакеты, ставить какие то интерсепторы. Научиться читать логи, запускаться в дебаге и что-то логировать?
Понимаю что мне никто ничем не обязан, но за спрос как говорится денег не берут.

**Comment by EvgenKo423** [comment]
TL;DR: Всё сложно.
Вообще по идее этим должны заниматься авторы, т.к. в россии DPI один и нет особого смысла делать разные настройки, достаточно нескольких рабочих пресетов с разными техниками, наиболее корректных / оптимальных с точки зрения работы сети.
Но, как многие уже говорили, у этого автора, к сожалению, с документированием проблемы (или со свободным временем). =(( Параметры не понятны не только вам, некоторые вещи я вообще в коде смотрел.

Но если хочется покопаться, то я вошёл так (правда у меня ИТшное образование и немного хакерского):
1. [Вот в этой портянке][zapret_docs] от господина **bol-van**'а очень хороший пересказ теории с разными нюансами. Там всё в куче и частично про его утилиты, так что фильтруйте (утилиты, кстати, тоже более продвинутые и вроде как даже лучшие для линуксов (**hufrea**, уж извини));
    - в частности, про [`--split`], [`--disorder`] и [`--fake`];
    - осложняется всё тем, что на уровне прокси это можно реализовать тока через костыли, поэтому если есть возможность, лучше использовать утилиты, работающие на уровне системы ([GoodbyeDPI] / nfqws из [zapret]'а, правда подключение последнего это отдельный вид извращений);
2. Запускаете мониторинг трафика (например, Wireshark'ом);
3. Загружаете сайт, отключаете захват и смотрите как должно выглядеть соединение;
4. Выбираете любой заблокированный сайт, желательно полегче и без поддоменов, и для упрощения анализа желательно чтобы в сеть в это время больше ничего не лезло;
    - разные сайты могут блокироваться по-разному, но в основном одинаково, по SNI;
5. Повторяете 2-3 и смотрите что происходит там;
    - в основном нас интересует Client Hello, в средней области отображается понятная структура пакета;
    - в хроме соединение с сайтом не закрывается, так что чтобы пошло новое соединение надо перед этим закрыть вкладку и на странице `chrome://net-internals/#sockets` нажать `Close idle sockets`;
6. Подбираете параметры, при которых в запросе будет использоваться разрешённый домен или DPI не сможет проверить запрос, а сервер его поймёт;
7. Повторяете 2-3 и смотрите что изменилось.

Но для `googlevideo.com` уже начинается такая магия, что даже не понятно как оно ещё может его проверять... Сейчас, судя по всему, корректно [туда пропускает][gv_now] только поддомены гугла, а не корректно поломки версии, всё с перепроверкой последнего пакета перед ответом сервера.

[zapret_docs]: https://github.com/bol-van/zapret/tree/master/docs
[`--split`]: https://github.com/bol-van/zapret/blob/243a64cbec5a7e4cdbace38789e65e5efe73ffdb/docs/readme.txt#L766-L782
[`--disorder`]: https://github.com/bol-van/zapret/blob/243a64cbec5a7e4cdbace38789e65e5efe73ffdb/docs/readme.txt#L838-L849
[`--fake`]: https://github.com/bol-van/zapret/blob/243a64cbec5a7e4cdbace38789e65e5efe73ffdb/docs/readme.txt#L297

[GoodbyeDPI]: https://github.com/ValdikSS/GoodbyeDPI
[zapret]: https://github.com/bol-van/zapret

[gv_now]: https://github.com/hufrea/byedpi/discussions/218#discussioncomment-11062421

**Comment by wassalam01** [comment]
Супер! Спасибо огромное за объяснение!

**Comment by zimbabwe303** [comment]
Моё скромное мнение.

Тут такая ситуевина: ТСПУ - это "чёрный ящик", поэтому сравнивая нормальный и покоцанный ClientHello не многое можно понять, по части того, какие настройки вставлять, ибо не ясно, на что именно реагирует эта самая вундервафля. По крайней мере мне эти сравнения практически ничего не подсказали (с другой стороны у меня двойка по сетевой магии).

Из опыта подборов понял, что самая полезная вещь не Wireshark, а банальный ключ `-x2 (--debug 2)` самого же byedpi. В его выдаче смотрим на строки, заканчивающиеся на `evt: EV_TUNNEL, mod_iter: [число]` - если они есть после попытки зайти на сайт, это говорит об успешном начале передаче пакетов (но не о скорости). Число `mod_iter` в норме должно быть около 400 или больше, маленькие числа говорят о том, что ТСПУ всё обнаружило и гадит в поток данных, чтобы его затормозить (желательно до нуля). С googlevideo.com дело обстоит именно так: ТСПУ коннект не разрывает, но гадит в него.

Ещё советы:

1. Не стоит тупо наворачивать кучу параметров не понимая, что делаешь. Осознанный минимализм всегда побеждает, а бессмысленный хаос - его и ТСПУ умеет создавать.
2. Большие значения имеющихся параметров редко выручают: если ТСПУ уже научено обнаружать тот или иной метод дурения, то размеры подставы существенного значения не имеют. В лучем случае в ТСПУ может сработать некий лимит на объём анализируемых данных, но, в таком случае, у вас всё равно передача будет тормозить.
3. Варианты дурения всегда стоит пробовать поменять на другие (`-s` на `-d`, `-q`, `-o`), а не тупо добавлять новые. Несколько подряд идущих однотипных методов могут помочь, но не более двух (как тот же популярный `-s1 -q1`), далее уже сам byedpi рискует не собрать всё обратно.
4. Пробуем добавлять модификаторы: `-Y (--drop-sack)`, `-S (--md5sig)`, `-F (--tfo)`. Они помогают, но не часто.
5. Нужно чутьё. Перебираем параметры, а в браузере перезагружаем страницу. Выдаётся уже другая ошибка? Значит копаем в правильном направлении и подобранные параметры лучше уже не менять, а двигаться от них: добавляем, для начала `-At`, и пробуем что-нибудь дальше, например `-f (--fake)` и/или `-r (--tlsrec)`.
6. Может помочь игра с размером буфера `-b (--buf-size)` и `-R (--round)`. Есть ещё более зверский вариант: добавляем в параметры сплита повторы и пропуски: например `-s1:3:1` (3 повтора, 1 пропуск).
7. Можно (и лучше) отталкиваться от уже найденных другими людьми минималистичных строк, то есть таких строк, в которых нет нагромождений типа `-s1 -s20 -q40 -o10` и в которых не более трёх `-A (--auto)`. Каждое `-A` (кроме заключительного `-An` или `-As`) срабатывает, чтобы пробиться через один этап защиты, а ТСПУ это, судя по всему, всего один. Дополнительные препоны это могут быть старая система Ревизор и подобная же фиговина у магистральщика, и они, как правило, перенаправляют на страницу с надписью "А низзя!" (c) Слава Полунин). На подобные перенаправления реагирует `-Ar (--auto r)`, либо тот же `-At (--auto t)` - смотря как это перенаправление организовано, и что срабатывает первым - ТСПУ или Ревизор.

**Comment by EvgenKo423** [comment]
> ТСПУ - это "чёрный ящик", поэтому сравнивая нормальный и покоцанный ClientHello не многое можно понять, по части того, какие настройки вставлять, ибо не ясно, на что именно реагирует эта самая вундервафля.

Ну да, тут надо думать как бы вы сами стали это блокировать, изобретать свой DPI.
После применения параметров в Wireshark'е можно увидеть к каким модификациям пакета они приводят, что доходит до сервера, а что нет. Ещё можно увидеть если параметры работают не для всех соединений с одним сайтом.

> С googlevideo.com дело обстоит именно так: ТСПУ коннект не разрывает, но гадит в него.

Из моих наблюдений, `googlevideo.com` уже давно блокируется сразу и полностью, просто там идёт параллельно несколько соединений и у него не всегда получается.

> 1\. Не стоит тупо наворачивать кучу параметров не понимая, что делаешь.

Добавлю, что это замедляет работу сети, особенно если нет фильтра по доменам.

> 6\. Может помочь игра с размером буфера `-b (--buf-size)`

С этим лучше не играть, этот параметр про оптимизацию копирования данных между прокси и ОС, на обход он не влияет.

> ... добавляем, для начала `-At` ...
> Каждое `-A` (кроме заключительного `-An` или `-As`) срабатывает, чтобы пробиться через один этап защиты, а ТСПУ это, судя по всему, всего один.

Вообще этот параметр предназначен для смены одного способа на другой: либо с общей строкой для всех сайтов, либо если по умолчанию не применяется ничего. Лично я собираю сайты с одинаковым способом обхода в один файл, а `--auto` использую на случай, если наткнусь на то, чего нет в списках. `-A n` указывается в конце блока, остальные в начале!

Несколько фильтров я на практике ни разу не видел, да и не понимаю я все эти упоминания про старые не выключенные системы: операторам тупо не выгодно держать включёнными несколько железок, особенно если они уже заменены на ТСПУ. Но, как говорится в описании параметров ByeDPI и zapret'а, в сети есть другие фильтры, для защита от DDoS и т.п. И чтобы пробиться через все "ящики", параметры должны удовлетворять их все одновременно, `--auto` тут не поможет (вроде).

> ... и они, как правило, перенаправляют на страницу с надписью "А низзя!"

Перенаправление может быть только по HTTP, но при правильной настройке наткнуться на него почти невозможно и обходить смысла нет:
* Почти все сайты используют HTTPS и поисковики тоже выдают ссылки с ним;
* При первом заходе на многие сайты по HTTPS они добавляются в специальную базу браузера, которая указывает, что в следующий раз к ним нужно обращаться только так. Эти записи хранятся годы и обновляются при посещении (и мешают тестить обходы...);
* В Хроме есть настройка "Всегда использовать безопасные соединения", при включении которой он будет всегда использовать HTTPS, но это может сломать сайты, которые хотят только HTTP.
  Или можно самому при первом заходе на новый сайт вводить в адресную строку `https://`. Раньше Хром использовал его по умолчанию, а с добавлением этой настройки как идиот стал использовать HTTP...

**Comment by vehametra** [comment]
```
TFO supported!
new conn: fd=13, addr=216.58.207.202:443
TFO supported!
new conn: fd=14, addr=216.58.207.202:443
host: youtubei.googleapis.com (515)
state: 2
split: pos=0-1 (1), m: DESYNC_SPLIT
send: Resource temporarily unavailable
send: Resource temporarily unavailable
split: pos=1-1 (-1), m: DESYNC_DISOOB
host: youtubei.googleapis.com (191)
state: 2
split: pos=0-1 (1), m: DESYNC_SPLIT
send: Resource temporarily unavailable
send: Resource temporarily unavailable
split: pos=1-1 (-1), m: DESYNC_DISOOB
offset: 1, skip
offset: 1, skip
offset: 1, skip
offset: 1, skip
close: fd=12 (pair=14), recv: 1143, rounds: 2
close: fd=14 (pair=12), recv: 384, rounds: 1
close: fd=9 (pair=10), recv: 7272, rounds: 6
close: fd=10 (pair=9), recv: 15807, rounds: 5
close: fd=7 (pair=8), recv: 715, rounds: 2
close: fd=8 (pair=7), recv: 258, rounds: 1
close: fd=11 (pair=13), recv: 6150, rounds: 3
close: fd=13 (pair=11), recv: 18947, rounds: 2
client_loop: send disconnect: Broken pipe
```
что значит round? Я могу в блоке -At  на 10й  попытке включить например опцию -M  и вторую --tls-rec?
_send: Resource temporarily unavailable_   Это нормальное поведение?
Что значит __state__?

**Comment by sakontwist** [comment]
round это видимо один обмен запрос-ответ. У меня стоят все дурения на --round 1 и этого хватает. Что касается auto, у меня переключается несколько режимов. Это просто для примера, реально может не работать:
```
--proto udp --udp-fake 5 --ttl 2 \
--auto=none --proto http --mod-http=h,d \
--auto=none --proto tls --disorder +s --split 0+sm --round 1 \
--auto=torst,redirect,ssl_err --proto tls --disorder 2 --split 7 --round 1
```

Видимо да, можно включить, например, --round 10-11, только смысла в этом немного

**Comment by radioxoma** [comment]
@EvgenKo423 поправьте ссылки с issue на discussions https://github.com/hufrea/byedpi/discussions/218#discussioncomment-11062421, в issue старый текст, не могу править.

**Comment by hufrea** [comment]
Для начала стоит ответить на другой вопрос. Как, перехватив первый пакет, понять на какой домен идет обращение?
Посмотрите как выглядит TLS ClentHello ([тут например](https://tls.dxdt.ru/tls.html#13-clienthello)).
По первым 6-ти байтам (или даже по 3-м, если грубо) можно понять, что это приветственное сообщение TLS.
Затем нужно перейти к блоку расширений, пропустив ненужное для нас. Делаем это так: 
Пропускаем первые 43 байта и читаем первый (относительно смещения, т.е. по позиции 43) байт. Этот байт - размер SessionID, добавляем к нему текущее смещение (44) и читаем уже 2 байта. Это - размер шифронаборов, добавляем его к текущей позиции. Пропускаем 2 байта, и вот мы у блока расширений. 
Заголовок расширения выглядит так - 2 байта тип + 2 байта размер. Домен хранится в расширении SNI, его тип 0x0000. И так, проходимся по блоку: читаем тип, если это не 0x0000, то добавляем к текущей позиции размер этого расширение, тем самым перейдя к следующему, снова читаем тип, и так пока не наткнемся на искомый.
Если знать на каком IP лежит сайт, однако сам IP заблокировать невозможно, т.к. на нем могут быть другие сайты, то можно поступить иначе. Можно ничего не парсить, а просто при обращении на этот адрес грубо искать домен в первых пакетах. Или же можно сделать список разрешенных доменов, что лежат на этом IP, и разрешать лишь те соединения, что их содержат.
Алгоритмы могут быть разные. Задача - понять и сломать их. На вооружении есть `split, disorder, fake, oob и tlsrec`. Что они делают написано в README, в конце есть дополнительные ссылки. Поняв, как работает DPI, вы также и поймете что и по каким позициям стоит применять. Для примерного понимания хватит и вышеописанного. Если хочется узнать, почему и как такие штуки, как split и disorder, работают, то стоит изучить еще и протокол TCP.

**Reply by EvgenKo423** [reply]
> Можно ничего не парсить, а просто при обращении на этот адрес грубо искать домен в первых пакетах

Только это сильно неэффективно по сравнению с парсингом. Странно, что они, судя по всему, к этому сейчас прибегают для гуглвидео, видимо сильно горит...

**Comment by hufrea** [comment]
> Но, как многие уже говорили, у этого автора, к сожалению, с документированием проблемы.

Было бы здорово, если бы говорили что именно написать, а лучше присылали бы PR. Мне сложно понять чего конкретно от меня хотят. Параметры описаны, в репозитории Zapret-а куча подробностей, не вижу смысла пересказывать их.

> Число mod_iter в норме должно быть около 400 или больше, маленькие числа говорят о том, что ТСПУ всё обнаружило и гадит в поток данных

Это число - номер итерации цикла событий, в который соединение было добавлено или изменено. Для первых соединений оно всегда маленькое.

> Может помочь игра с размером буфера -b (--buf-size)

Я уже писал - этого делать не стоит. Это скажется на скорости соединения и потреблении ресурсов процессора (чем меньше размер буфера, тем больше циклов чтения будет совершенно). Если параметр `-b` с определенным числом влияет на обход, то стоит поиграться со сплитами на позициях кратными этому числу.

> Пробуем добавлять модификаторы: -Y (--drop-sack), -S (--md5sig), -F (--tfo)

`-Y` вообще не стоит использовать. Он добавляет существенную задержку к соединению (около 200ms) и ломает disorder. Он имеет смысл только при особом [сочетании параметров ](https://github.com/hufrea/byedpi/blob/75671fa11c57aa458e87d876efc48b40985ed78f/README.md?plain=1#L207-L212). 
На `--tfo` опираться не следует, т.к. большинство сайтов его не поддерживают. При подборе он только внесет неясность.

**Reply by vehametra** [reply]
> . Если параметр -b с определенным числом влияет на обход, то стоит поиграться со сплитами на позициях кратными этому числу.

 вот это здорово.
 
>     Было бы здорово, если бы говорили что именно написать, а лучше присылали бы PR. 
Я пытался разобраться в исходном тексте, чтобы  разобраться в выводе. И только больше запутался. Правда это было месяца 1,5 назад. Мне лично нехватает коментрариев в коде, которые бы облегчили понимание лога.

**Reply by EvgenKo423** [reply]
> Было бы здорово, если бы говорили что именно написать, а лучше присылали бы PR. Мне сложно понять чего конкретно от меня хотят. Параметры описаны ...

Не особо, есть пробелы. Напишу, что (было) не понятно мне:
- дурилки в целом: 
   - от какого места начинается отсчёт позиции? (да, можно проверить)
     (ответ: `--tlsrec`: TLS data / TCP data + 5, остальные: SSL / TCP data);
   - для многих, похоже, не очевидно, что указывать одинаковые номера позиций для разных опций нельзя (за исключением `--tlsrec`);
   - > На практике оптимально использовать:

       Обычно не понятно в чём заключается оптимальность. Например, когда у GoodbyeDPI написано, что `-e 32` оптимальней, чем `-e 2`, то это хотя бы с ИТшными знаниями можно как-то обосновать...;

- `--split`:
  `+e` раньше было не описано, я нашёл его случайно в коде и в каком коммите оно было добавлено можно было вычислить только перебором.
  > Позиция имеет вид offset[:repeats:skip]

  Двусмысленно: 1:3:5 можно истолковать как отрезать 1, продублировать его 3 раза, следующие 5 байт не отправлять (окей, есть пример);

- `--disorder`: не понятно, почему со сплитом ретрансмиссия становится не полная, я не поверил, пока сам не увидел. Стоит упомянуть, что в связи с особенностями работы ОС ретрансмиссия начинается с позиции, с которой начались потери (максимальный ACK, полученный от сервера);

- `--fake`:
  > Количество байт отправляемого из фейка равно рамеру разбиваемой части

  Количество данных не может отличаться от размера оригинального запроса: если больше - обрезается, если меньше - добивается нулями (при значениях `--fake`, близких к концу запроса);
- `--fake-offset`: может быть не понятно относительно чего смещение. Насколько я понял, указать можно только 1 раз для группы, следовательно (корректных) фейков на группу может быть только 1;
- `--fake-data`:
  по умолчанию используется оригинальный запрос? (ответ: нет)
  формат (как в консоли указать двоичные данные)? (ответ: `\xHH`);
- `--oob-data <char>`: можно двоичный? (ответ: да);
- `--tls-sni`: в `--fake-data` тоже меняет? (ответ: нет);

- `--auto`: [можно указывать несколько одинаковых](/hufrea/byedpi/discussions/224#discussioncomment-11198878) (кроме `none`)? многие походу так делают (без ограничений)...;
   - `torst`: написано, что группы обходятся по порядку, но при отвале по таймауту, вроде как, сразу идёт в эту группу, а не в следующую по порядку;
   - `none`: описание вообще не понятно. Добавьте людям пример с несколькими одинаковыми группами;

- `--ip-opt`: формат? на винде не то что _может_ быть отключена (кем?), а точно исключена;

- `--auto-mode`: не понятно как / зачем использовать;
- `--cache-ttl`: время, на которое запоминается сработавшая группа параметров для IP? в основном актуально для авто, видимо;
- `--timeout`: по умолчанию его нет, правильно?;
- `--round`: тупой вопрос от ИТшника - с 1 или с 0? =));
   - какие дурилки применимы к последующим кругам? фейк, полагаю, нет?;

- `--no-domain`: двусмысленно, т.к. пользователи к сайтам по домену и подключаются. Стоит упомянуть, что это такая фича SOCKS (которую фиг пойми где по умолчанию используют);

Как видите, если погружаться, то больше вопросов, чем ответов. Без выкуривания ящика пол литров тут не разобраться.

> `-Y` вообще не стоит использовать. Он добавляет существенную задержку к соединению (около 200ms) и ломает disorder. Он имеет смысл только при особом [сочетании параметров ](https://github.com/hufrea/byedpi/blob/75671fa11c57aa458e87d876efc48b40985ed78f/README.md?plain=1#L207-L212).

Тоже будет полезно в ридми. ;-)

**Reply by sakontwist** [reply]
Группа "--auto none ..." применяется, если предыдущая группа не применилась к запросу. Например, там было жесткое условие вроде "--pf 80" или "--proto udp" или "--hosts" и запрос под него не попал.

Можно делать несколько --auto, если в каждом определены условия, по которым сработает none.

--auto-mode 1 запоминает группу --auto, которая сработала при обращении к данному ip, это сопровождается в логе записью "save ip", при следующем обращении к адресу применяется сохраненная группа, в логе будет запись "m=<номер группы>". Первая группа (до первого auto) имеет номер 0.

--round начинается с 1, это видно в логе.

--no-domain у меня вообще не работает ) Я направляю на порт сервиса определенные запросы через routing xray и эта опция все ломает....

**Reply by EvgenKo423** [reply]
@sakontwist
> Можно делать несколько --auto, если в каждом определены условия

Т.е. можно сделать несколько `--auto=torst` с ограничением по доменам?

**Reply by sakontwist** [reply]
Обычно я использую такие конструкции:
```
--proto udp --udp-fake 5 --ttl 2 \   # m=0
--auto=none --proto http --mod-http=h,d \   # m=1
--auto=none --proto tls --disorder 2 --split 7 \ # m=2
--auto=torst,ssl_err --proto tls --disorder 0+se  # m=3
```
с одним условием для rst/err, но это всего лишь условие, которое срабатывает на ошибку. Согласно доке, проверка выполняется по порядку: "Сначала проверяется триггер, указанный в auto, затем pf, proto и hosts". Получается, что добавление в группе "--auto=torst" условия --hosts применит правила только в случае совпадения хоста, но вот что будет при добавлении еще одного условия далее - трудно сказать (теоретически оно сработает только при --auto=none). Как наследуется по группам признак "ошибки" знает только автор...

**Reply by radioxoma** [reply]
`--timeout`: по умолчанию не включён. Включается для всех групп одновременно. В итоге использую `ciadpi $PARAMS -T3 -At`.

**Comment by radioxoma** [comment]
@hufrea сконвертируйте, пожалуйста, issue в discussion, я добавлю туда листинг конфига для колорайзера [grc](https://github.com/garabik/grc). Не хочется сразу выкладывать т..к. отключится возможность редактирования.

**Comment by hufrea** [comment]
> сконвертируйте, пожалуйста, issue в discussion

Хорошо.
> я добавлю туда листинг конфига для колорайзера [grc](https://github.com/garabik/grc).

А разве это имеет какое-то отношение к заданному вопросу?

**Reply by vehametra** [reply]
К подбору конфига? ДА КОНЕЧНО. Единствено, что  раскрасна не вполне полна. Я так понимаю нужно учитывать значение в mod_iter?

**Reply by radioxoma** [reply]
Подсветил `mod_iter` красным, если число трёхзначное (ограничение regexp, не могу выставить условие вроде `<400`).

**Comment by radioxoma** [comment]
> А разве это имеет какое-то отношение к заданному вопросу?

Я в сетях и протоколах не разбираюсь, в логах не понятно, какие события хорошие, а какие плохие. Поэтому раскрасил. Особенно облегчает подбор параметров, если есть множественные `-T3 -At`. Легче заметить, если какой-то домен не разблокировался или сломался. Решил, что на PR мой вклад не тянет.

**Конфиг колорайзера [grc](https://github.com/garabik/grc) для облегчения чтения логов byedpi, понимания TCP Flow и, соответственно, подбора параметров обхода.** Блокировки подсвечиваются как `blink on_red`. Сохранить конфиг в файл `conf.ciadpi` и запустить в том же каталоге:

```sh
# grc config for https://github.com/hufrea/byedpi
# grc -ec conf.ciadpi ciadpi -p1081 -x1
# https://github.com/hufrea/byedpi/discussions/224
-
##
## Colorize by group index
##
regexp=^desync TCP.*
colour=black on_green
-
regexp=^host.*
colour=black on_white
-
regexp=^not resolved:.*
colour=red
-
##
## Colorize `DESYNC_.*` attacks
##
regexp=^split.*(DESYNC_.*)
colour=default, black on_green
-
# The only safe DESYNC without background
regexp=^split.*(DESYNC_SPLIT.*)
colour=default, green
-
regexp=^tlsrec
colour=on_blue
-
##
## Colorize close event
##
regexp=^close:.*
colour=red
-
# DPI filter triggered?
# Highlight zero data transfer (connection may be blocked by DPI)
regexp=^close:.*rounds: 0
colour=blink on_red
-
# recv: Connection timed out
# sendto: Resource temporarily unavailable
regexp=^recv.*|^sendto.*
colour=blink red
-
# DPI filter or config breaks site
# recv: Connection reset by peer (connection may be blocked by DPI)
regexp=^recv: Connection reset.*
colour=blink on_red
-
##
## Colorize various messages
##
regexp=^TFO supported!
colour=dark green
-
regexp=^pollout.*
colour=dark green
-
# Highlight Cloudflare canary domain for ECH
# https://blog.cloudflare.com/announcing-encrypted-client-hello
regexp=^host: cloudflare-ech.com
colour=green on_white
-
# -x2 logs: darken 
regexp=^new event.*
colour=dark
-
# -x2 logs: low values (<400) means DPI throttle?
regexp=.*EV_TUNNEL.*mod_iter: \d{1,3}$
colour=red
-
```
![Снимок экрана_2024-11-07_12-19-17](https://github.com/user-attachments/assets/cb322540-8fd6-45df-8d0d-8e529b5576ff)

**Reply by vehametra** [reply]
![Снимок экрана_2024-11-09_02-11-19](https://github.com/user-attachments/assets/e6d4a1a0-d6b6-433a-98e6-3a0a68f79348)
 У меня  `desync TCP`  вместо `desync params index:`    я так понимаю ,  m - это  param index, а r -  раунд

**Reply by vehametra** [reply]
скорее  `regexp=^desync TCP, m=([0-9]+), r=([0-9]+)`

**Reply by vehametra** [reply]
и  pollout  это хорошо. Как я понимаю, это значит что данные для клиента готовы. но это не единственное сообщение связаное с успешным пробоем. Даже не обязательно. и появляется только когда всё очень хорошо

**Reply by radioxoma** [reply]
> У меня `desync TCP` вместо `desync params index:` я так понимаю , m - это param index, а r - раунд

Обновил конфиг для  v0.15.

> recv: Connection reset by peer
> close: fd=6 (pair=5), recv: 0, rounds: 0  # Нули - плохо

Отсутствие _Connection reset_ и _rounds_ > 0 (они оба на красном фоне) говорят об успешном соединении. А как _pollout_ выглядит? Можно выделить зелёным.

**Reply by vehametra** [reply]
> > У меня `desync TCP` вместо `desync params index:` я так понимаю , m - это param index, а r - раунд
> 
> Обновил конфиг для v0.15.
> 
> > recv: Connection reset by peer
> > close: fd=6 (pair=5), recv: 0, rounds: 0  # Нули - плохо
> 
> Отсутствие _Connection reset_ и _rounds_ > 0 (они оба на красном фоне) говорят об успешном соединении. А как _pollout_ выглядит? Можно выделить зелёным.

```
TFO supported!
new conn: fd=70, addr=142.251.1.139:443
host: s.youtube.com (355)
tlsrec: pos=356, n=985
state: 2
split: pos=0-1 (1), m: DESYNC_SPLIT
state: 2
split: pos=1-2 (-1), m: DESYNC_SPLIT
offset: 1, skip
split: pos=0-2 (2), m: DESYNC_SPLIT
not sent after 0 ms
split: pos=2-3 (1), m: DESYNC_SPLIT
offset: 1, skip
send: pos=3-984
close: fd=69 (pair=70), recv: 985, rounds: 1
close: fd=70 (pair=69), recv: 374, rounds: 1
_send: 808 != 1400 (fd: 47)_
**pollout (fd=47)**
_send: 168 != 1400 (fd: 47)_
**pollout (fd=47)**
```
я выделил, но немножко налажал с регуляркой, посему не показвыаю.
Как я понял, pollout  возникает, когда  данные отправить клиенту нужно срочно, чтобы сразу освободиь буффер.
Но это не точно.
курсивом вроде send  данных клиенту,  обрати внимание на `fd`

![Снимок экрана_2024-11-10_03-41-18](https://github.com/user-attachments/assets/09f50640-9f57-4509-b091-d70e7767a794)

**Reply by radioxoma** [reply]
Никогда не видел pollout в своих логах. Выделил `colour=dark green`.

**Reply by vehametra** [reply]
> Никогда не видел pollout в своих логах. Выделил `colour=dark green`.

а `send:  ...!= ... (fd: ..)`  ?
А грипни лог.

**Comment by vehametra** [comment]
У меня по отдельности ни один набор опций не работает хорошо и долго.
![Снимок экрана_2024-11-09_02-01-49](https://github.com/user-attachments/assets/b04a96ae-26cd-46b1-852f-79b0f2589565)
 А вместе всё неплохо. Извините что скрином.  Переключение метода запутывания траффика помогает. Пригождаются даже довольно слабые варианты. важный  момент, резко повышающий эффективность  --tls-rec  разрезающий  sni  или хост. _От границы ключ переломлен пополам..._

**Reply by vehametra** [reply]
опция --md5-sig хорошо  сработала глубокой ночью. ранее с ней были проблемы на роутере с  OpenWRT. Сейчас вроде работает.
`# echo 1 > /proc/sys/net/ipv4/tcp_syncookies`
Это ведь оно?

**Comment by Axa-Ru** [comment]
Мой провадер ростелек.
Запускаю с параметрами автора

`ciadpi --hosts "wl_ciadpi.txt" -An --disorder 1 --auto=torst --tlsrec 1+s `

Файл wl_ciadpi.txt
```

adfox.ru
adriver.ru
aestatic.net
aliclick.shop
aliexpress.ru
alicdn.com
aliprice.com
avito.ru
b2c.passport.rt.ru
balsat.ru
googleapis.com
gosuslugi.ru
habr.com
idriver.ru
isdelivr.net
lastpass.com
lemanapro.ru
mail.ru
mosoblgaz.ru
mos.ru
obd-memorial.ru
online.vtb.ru
ozon.ru
shop.orionspb.ru
skcrtxr.com
sttsmntz.ru
tbank.ru
top-fwz1.mail.ru
urban-university.pro
vtb.ru
yandex.ru
```

**Comment by bombom-cmd** [comment]
Доброго всем дня! Недавно скачал себе byeDPI на ПК, так как zapret перестал работать с Дискордом. Почитал мануал где идет описание аргументов, всё очень сложно.
Нашел старую, уже нерабочую стратегию где был использован данный аргумент
```
-l, --fake-data <file|:str>
    Указать свои поддельные пакеты
    Строка может содержать escape символы (\n,\0,\0x10)
```
Вот сама страта
`-Ku -l:\xC2\x00\x00\x00\x01\x14\x2E\xE3\xE3\x5F\x6B\xBB\x23\xA8\xE6\x5D\xA9\x78\x21\xCF\xC2\x72\x4C\x8F\xC4\x5E\x14\x00\x00\x00\x00\xC5\x00\x00\x00\x00\x4C\x00\xA7\x00\x00\x00\x00\x00\x00\x44\x00\x00\x80\x00\x00\x00\x0D\xFC\xFA\x1D\xCD\x73\xBA\x2A\x90\x93\xB3\xEE\xF7\x43\xC5\x85\xDA\xFF\x45\x3C\x00\x00\x00\x00\x00\x00\x7C\x00\x9B\x00\xF6\x00\x00\xDD\x00\x00\x00\x00\x00\x00\x00\x00\x00\x59\xA8\xE4\x00\x00\x00\x00\x00\x00\x00\x00\x7B\x00\x0F\x00\x00\x00\x48\x4E\x00\x00\x00\x06\xF3\x00\x00\x00\x00\xD9\x5A\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00 -a3 -An --hosts list.txt -o1 -d1 -r1+s -t10 -b4000 -s1+s -s3+s -s6+s -s9+s -s12+s -s15+s -s20+s -s30+s -As --hosts list.txt -q1+s -s29+s -o5+s -f3 -S -As --hosts list.txt -d1+s -s3+s -d5+s -s7+s -r10+s -Mh,d -An`
Собственно в ней после аргумента **-l** идет нагромождение символов (в моем понимании), которое я не смог понять (что это за кодировка, в каком формате и чего собственно). Собственно к другим аргументам у меня тоже вопросы, но там хотя бы цифр и букв после них идет меньше и шансов ошибиться соответственно тоже ( я просто планирую как обезьяна повторить это в своей стратке меняя цифры на +-1 в одну из сторон и буквы также), но вот с аргументом -l вообще непонятно, что впринципе туда писать? 
Буду рад любой помощи. Заранее спасибо!

**Reply by EvgenKo423** [reply]
База описана в zapret'е и в начале темы. Аргумент `-l`, судя по всему, нужен для голосовых чатов, но, судя по качеству других параметров, смысла в его содержимом тоже может не быть. Писать туда что-то "как обезьяна" смысла нет, для этого, полагаю, пойдёт и пакет по умолчанию.

Но вообще проблема не в zapret'е, а в способе дурения. zapret – вещь более продвинутая, нет смысла менять утилиты, стоит сосредоточиться на понимании какой-то одной и методов дурения.
